### DISCUSSION TITLE
Caprover - NGINX Configuration

### DISCUSSION THREAD

**Discussion by wirpo032** [caprover/caprover#1460] [Q&A]
Hi, 

I use Caprover to deploy (among others things) a couple of static sites, each having their own domain. Everytime I do any change, the whole Caprover NGINX breaks (none of the websites including the server dashboard are accessible). I want to perform two amendments to the current config.

1. I was wondering to adjust the NGINX Configuration to include the following: make subdomain.domain.com show the website domain.com/subdomain. I found the following NGINX configuration snippet, that should accomplish that:

```
server {
    server_name blog.example.com;
    return 301 http://www.example.com/blog$request_uri;
}
```

2. I also want to add autoindexing for each folder without an index.html/index.htm. I found the following NGINX configuration.

```
location / {
    autoindex on;
}
```

When I head to the caprover server dashboard > website name > NGINX Configuration, I find the following configuration. I wanted to insert the code snippets in the following spots (marked with ** here below).

```
<%
if (s.forceSsl) {
%>
    server {

        listen       80;

        server_name  <%-s.publicDomain%>;

        **server_name blog.example.com;        
        return 301 http://www.example.com/blog$request_uri;**

        # Used by Lets Encrypt
        location /.well-known/acme-challenge/ {
            root <%-s.staticWebRoot%>;
        }

        # Used by CapRover for health check
        location /.well-known/captain-identifier {
            root <%-s.staticWebRoot%>;
        }

        location / {
            return 302 https://$http_host$request_uri;
            **autoindex on;**
        }
    }
<%
}
%>


server {

    <%
    if (!s.forceSsl) {
    %>
        listen       80;
    <%
    }
    if (s.hasSsl) {
    %>
        listen              443 ssl http2;
        ssl_certificate     <%-s.crtPath%>;
        ssl_certificate_key <%-s.keyPath%>;
    <%
    }
    %>

        client_max_body_size 500m;

        server_name  <%-s.publicDomain%>;

        # 127.0.0.11 is DNS set up by Docker, see:
        # https://docs.docker.com/engine/userguide/networking/configure-dns/
        # https://github.com/moby/moby/issues/20026
        resolver 127.0.0.11 valid=10s;
        # IMPORTANT!! If you are here from an old thread to set a custom port, you do not need to modify this port manually here!!
        # Simply change the Container HTTP Port from the dashboard HTTP panel
        set $upstream http://<%-s.localDomain%>:<%-s.containerHttpPort%>;

        location / {

    <%
    if (s.httpBasicAuthPath) {
    %>
            auth_basic           "Restricted Access";
            auth_basic_user_file <%-s.httpBasicAuthPath%>; 
    <%
    }
    %>

            proxy_pass $upstream;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;

    <%
    if (s.websocketSupport) {
    %>
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_http_version 1.1;
    <%
    }
    %>
        }

        # Used by Lets Encrypt
        location /.well-known/acme-challenge/ {
            root <%-s.staticWebRoot%>;
        }
        
        # Used by CapRover for health check
        location /.well-known/captain-identifier {
            root <%-s.staticWebRoot%>;
        }

        error_page 502 /captain_502_custom_error_page.html;
        location = /captain_502_custom_error_page.html {
                root <%-s.customErrorPagesDirectory%>;
                internal;
        }
}
```

I'm confident that the way I do it is wrong, but I do not know the correct way to solve this issue. Can anyone help?

TIA

**Comment by githubsaturn** [comment]
Not sure what you're trying to do.... Are you trying to redirect blog.domain.com to domain.com/blog? 
Or do you want to rewrite it?


In simple terms, what do you expect to happen when you type blog.domain.com:
* Should it redirect the URL to domain.com/blog and load this URL?
* Or should it simply load blog.domain.com, but behind the scenes it fetches the content from domain.com/blog

**Comment by wirpo032** [comment]
The second option you mentioned:

> Or should it simply load blog.domain.com, but behind the scenes it fetches the content from domain.com/blog

Is this feasible?

**Reply by githubsaturn** [reply]
It's feasible but you'll need to create a separate app, or at least a separate nginx server block for it. Right now, you have two server_name statements in a single server block and that's what breaks your nginx.
